lib/engine-object-file/src/serialize.rs

```diff
@@ -20,19 +46,55 @@ pub struct ModuleMetadataSymbolRegistry {
 }
 
 impl ModuleMetadata {
-    pub fn split(&mut self) -> (&mut CompileModuleInfo, ModuleMetadataSymbolRegistry) {
-        let compile_info = &mut self.compile_info;
-        let symbol_registry = ModuleMetadataSymbolRegistry {
-            prefix: self.prefix.clone(),
-        };
-        (compile_info, symbol_registry)
-    }
-
     pub fn get_symbol_registry(&self) -> ModuleMetadataSymbolRegistry {
         ModuleMetadataSymbolRegistry {
             prefix: self.prefix.clone(),
         }
     }
+
+    pub fn serialize(&self) -> Result<Vec<u8>, CompileError> {
+        let mut serializer = SharedSerializerAdapter::new(WriteSerializer::new(vec![]));
+        let pos = serializer.serialize_value(self).map_err(to_compile_error)? as u64;
+        let mut serialized_data = serializer.into_inner().into_inner();
+        serialized_data.extend_from_slice(&pos.to_le_bytes());
+        if cfg!(target_endian = "big") {
+            serialized_data.extend_from_slice(&[b'b']);
+        } else if cfg!(target_endian = "little") {
+            serialized_data.extend_from_slice(&[b'l']);
+        }
+        Ok(serialized_data)
+    }
+
+    pub unsafe fn deserialize(metadata_slice: &[u8]) -> Result<Self, DeserializeError> {
+        let archived = Self::archive_from_slice(metadata_slice)?;
+        Self::deserialize_from_archive(archived)
+    }
+
+    unsafe fn archive_from_slice<'a>(
+        metadata_slice: &'a [u8],
+    ) -> Result<&'a ArchivedModuleMetadata, DeserializeError> {
+        let mut pos: [u8; 8] = Default::default();
+        let endian = metadata_slice[metadata_slice.len() - 1];
+        if (cfg!(target_endian = "big") && endian == b'l')
+            || (cfg!(target_endian = "little") && endian == b'b')
+        {
+            return Err(DeserializeError::Incompatible("incompatible endian".into()));
+        }
+        pos.copy_from_slice(&metadata_slice[metadata_slice.len() - 9..metadata_slice.len() - 1]);
```

Comment:
The last eight bytes? What if `metadata_slice` is too short?

Reply:
Fixed!